testing.md

```diff
@@ -34,6 +34,82 @@ A piece of code is not done until it's tested.
+environment.  As an example, if to test your code, you bring up an instance of your service and
+manually send it an example request, then this flow should be encoded as an integration test.
+
+### Test Deterministically
+
+Tests should be deterministic. That means that if you run a test multiple times without changing the
+code, it should produce the same results. A test that sometimes fails and sometimes passes is not
+consistent, and therefore can't be relied on to give you confidence about the correctness of a
+program.
+
+Concretely, this means you should avoid relying on factors like randomness, time, or external
+services in the structure of your tests. For instance, if you are testing a button that changes a
+piece of UI, you shouldn't wait for an arbitrary number of seconds to check that the UI refreshes,
+since the time that it takes to refresh is not deterministic. Instead, you can emit some sort of
+event when the UI is finished refreshing, and make the test listen for this event.
+
+Testing non-deterministic code can also be very difficult. If you can, avoid non-deterministic
+behavior in your code. If you must have it, try to isolate it so it can be deterministically mocked
+in tests.
+
+One type of non-determinism that *is* useful for testing is generating random inputs.
+[Generative testing](http://blog.8thlight.com/connor-mendenhall/2013/10/31/check-your-work.html), in
+which a program generates different test inputs each time the tests are run, can help describe
+properties of your program that should hold for all inputs. They can make a stronger assertion about
+correctness, since, if they pass, it means your program works for many random inputs, not just a few
+representative ones. Generative tests can also come up with edge cases that you may not have thought
+of.
+
+### Prefer Readability and Ease of Code
+
+Tests should be written in a manner that prioritizes readability and ease of writing new tests and
+modifying existing ones. In particular, these should be prioritized over other typical concerns like
+abstraction and removing code duplication. If abstractions are needed, they should be created in a
+manner that preserves the logic in the test functions code without requiring many hops. Another way
+of saying this is that it should be apparent what the test is doing by just looking at the test
+code.
+
+As an example, we had a testing library called which populated a bunch of fake data in the database
+specified by the test. A lot of our tests used with a single statement of the form
+`db_test_helpers.set_up_test_data`. This single function populated a bunch of state within the
+database including creating students, teachers, districts, apps. However, it also hid away all the
+details of the state being created. This meant that it became very hard to reason about tests or
+modify your tests if the current state was not appropriate for them.
```

Comment:
in my experience, the main problem with having the shared fake data was that modifying the fake data impacted all of the dependent tests, not that the data was hidden. so i'm not sure this example is completely relevant to the point you're making here. is there another example you can think of?